pkg/api/ldap_debug.go

```diff
@@ -141,6 +153,77 @@ func (server *HTTPServer) GetLDAPStatus(c *models.ReqContext) Response {
 	return JSON(http.StatusOK, serverDTOs)
 }
 
+// PostSyncUserWithLDAP enables a single Grafana user to be synchronized against LDAP
+func (server *HTTPServer) PostSyncUserWithLDAP(c *models.ReqContext) Response {
+	if !ldap.IsEnabled() {
+		return Error(http.StatusBadRequest, "LDAP is not enabled", nil)
+	}
+
+	ldapConfig, err := getLDAPConfig()
+
+	if err != nil {
+		return Error(http.StatusBadRequest, "Failed to obtain the LDAP configuration. Please verify the configuration and try again", err)
+	}
+
+	userId := c.ParamsInt64(":id")
+
+	query := models.GetUserByIdQuery{Id: userId}
+
+	if err := bus.Dispatch(&query); err != nil { // validate the userId exists
+		if err == models.ErrUserNotFound {
+			return Error(404, models.ErrUserNotFound.Error(), nil)
+		}
+
+		return Error(500, "Failed to get user", err)
+	}
+
+	authModuleQuery := &models.GetAuthInfoQuery{UserId: query.Result.Id, AuthModule: models.AuthModuleLDAP}
+
+	if err := bus.Dispatch(authModuleQuery); err != nil { // validate the userId comes from LDAP
+		if err == models.ErrUserNotFound {
+			return Error(404, models.ErrUserNotFound.Error(), nil)
+		}
+
+		return Error(500, "Failed to get user", err)
+	}
+
+	ldapServer := newLDAP(ldapConfig.Servers)
+	user, _, err := ldapServer.User(query.Result.Login)
+
+	if err != nil {
+		if err == ldap.ErrCouldNotFindUser { // User was not in the LDAP server - we need to take action:
+
+			if setting.AdminUser == query.Result.Login { // User is *the* Grafana Admin. We cannot disable it.
+				errMsg := fmt.Sprintf(`Refusing to sync grafana super admin "%s" - it would be disabled`, query.Result.Login)
+				logger.Error(errMsg)
+				return Error(http.StatusBadRequest, errMsg, err)
+			}
+
+			// Since the user was not in the LDAP server. Let's disable it.
+			login.DisableExternalUser(query.Result.Login)
+			return JSON(http.StatusOK, "User disabled without any updates in the information") // should this be a success?
+		}
+	}
+
+	upsertCmd := &models.UpsertUserCommand{
+		ExternalUser:  user,
+		SignupAllowed: setting.LDAPAllowSignup,
+	}
+
+	err = bus.Dispatch(upsertCmd)
+
+	if err != nil {
+		return Error(http.StatusInternalServerError, "Failed to udpate the user", err)
+	}
+
+	err = tokenService.RevokeAllUserTokens(context.TODO(), userId)
```

Comment:
Not sure I understand this correctly - why do we remove tokens for users that found in LDAP, but don't do this for removed users?